1) Load the Titanic dataset using Pandas. Create a Series from the 'Age' column. Examine the distribution of passenger ages using descriptive statistics and identify the age group with the highest fatality rate. Justify your observations.

i) Loading the dataset from kaggle

In [1]:
import pandas as pd

In [4]:
df = pd.read_csv('C:/Users/Praneeth Reddy/Downloads/titanic.csv')
df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


We are loading the dataset from kaggle and importing it into the notebook using pd.read_csv('path').
df.head() is used to display the first five rows of the dataset.

ii) Creating a series from Age column

In [8]:
tit_age = df['Age']
tit_age

0      22.0
1      38.0
2      26.0
3      35.0
4      35.0
       ... 
886    27.0
887    19.0
888     NaN
889    26.0
890    32.0
Name: Age, Length: 891, dtype: float64

iii) Descriptive statistics 

In [9]:
tit_age.describe()

count    714.000000
mean      29.699118
std       14.526497
min        0.420000
25%       20.125000
50%       28.000000
75%       38.000000
max       80.000000
Name: Age, dtype: float64

describe() shows the current descriptive statistics of all the columns in the dataset.

iv) AgeGroup with highest fatality rate

In [14]:
bins = [0, 12, 18, 60, 100]
labels = ['Child', 'Teen', 'Adult', 'Senior']
df['AgeGroup'] = pd.cut(df['Age'], bins=bins, labels=labels)

fatality_rate = 1 - df.groupby('AgeGroup')['Survived'].mean()
fatality_rate

AgeGroup
Child     0.420290
Teen      0.571429
Adult     0.611212
Senior    0.772727
Name: Survived, dtype: float64

To analyze survival rates across different life stages the 'Age' variable is divided into categorical buckets.
'bins' are used to define intervals/ranges of age, instead of using raw numbers like 25, 30 etc, ages are grouped into categories.
'labels' are names given to each bin so that it becomes easy for us to read and understand.
pd.cut() converts continuous/raw data to categorial data.

Categorical data is way easier to use in groupby() operations. The final output will clearly say if its a 'Child' or a 'Senior' instead of giving confusing intervals.

df.groupby('AgeGroup')['Survived'].mean() will give the 'Survived' rate and 1 - 'Survived' rate will give us 'Fatality' rate.

In [15]:
print(f"The group with the highest fatality rate is: {fatality_rate.idxmax()}")

The group with the highest fatality rate is: Senior


While researching I've found out that idxmax() is used to find the index label of the first occurrence of the maximum value in a Series or DataFrame.

From the output we can clearly see that 'Senior' and 'Adult' age groups have the highest fatality rate.
This can be justified as cruises usually follow 'Women and Children First' policy where they are given access to lifeboats first. 
We can see that the mortality rate for children is less.
On the other hand, the mortality rate for Senior group is more as they were less likely to reach the lifeboats in the chaos.

2) Using the DataFrame, determine which passenger class (Pclass) had the best survival rate. Apply boolean indexing and conditional selection to filter survivors and non-survivors, then compare results across genders.

i) Passenger class survival rate

In [16]:
df.groupby('Pclass')['Survived'].mean()

Pclass
1    0.629630
2    0.472826
3    0.242363
Name: Survived, dtype: float64

From the output we can see that P1 class has the highest survival rate. This is due to socioeconomic priority and lifeboat access.

ii) Boolean indexing and condition selection

In [17]:
survivors = df[df['Survived'] == 1] 
non_survivors = df[df['Survived'] == 0]

We are filtering survivors and non-survivors using Boolean Indexing.

In [23]:
survivor_gender_counts = survivors['Sex'].value_counts()
survivor_gender_counts

female    233
male      109
Name: Sex, dtype: int64

In [24]:
non_survivor_gender_counts = non_survivors['Sex'].value_counts()
non_survivor_gender_counts

male      468
female     81
Name: Sex, dtype: int64

We are using Conditional Selection to compare results across genders. 
value_counts() is a method used to count the unique occurrences of values in a Series.

In [18]:
df.groupby(['Pclass','Sex'])['Survived'].mean()

Pclass  Sex   
1       female    0.968085
        male      0.368852
2       female    0.921053
        male      0.157407
3       female    0.500000
        male      0.135447
Name: Survived, dtype: float64

In [27]:
female_survival = df[df['Sex'] == 'female']['Survived'].mean()
male_survival = df[df['Sex'] == 'male']['Survived'].mean()

print(f"Female Survival Rate: {female_survival:.2%}")
print(f"Male Survival Rate: {male_survival:.2%}")

Female Survival Rate: 74.20%
Male Survival Rate: 18.89%


To compare results more effectively, Survival Rates across genders is calculated. 

Even though P1 class had the best survival rate due to socioeconomic priority, gender was an even stronger predictor.
Boolean indexing reveals that a vast majority of survivors were female ( 'Women and Children First' policy ), while the majority of non-survivors were male.

3) Identify all missing values in the dataset. Compare the effect of replacing missing 'Age' values with the mean versus the median. Which strategy is more appropriate and why? Support your answer with statistical evidence.

i) Identifying missing values

In [28]:
df.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
AgeGroup       177
dtype: int64

ii) Mean vs Median

In [42]:
mean_val = df['Age'].mean()
median_val = df['Age'].median()

In [43]:
mean_val

29.69911764705882

In [44]:
median_val

28.0

In [45]:
df_mean = df['Age'].fillna(mean_val)
df_mean

0      22.000000
1      38.000000
2      26.000000
3      35.000000
4      35.000000
         ...    
886    27.000000
887    19.000000
888    29.699118
889    26.000000
890    32.000000
Name: Age, Length: 891, dtype: float64

In [46]:
df_median = df['Age'].fillna(median_val)
df_median

0      22.0
1      38.0
2      26.0
3      35.0
4      35.0
       ... 
886    27.0
887    19.0
888    28.0
889    26.0
890    32.0
Name: Age, Length: 891, dtype: float64

In [47]:
original_stats = df['Age'].describe()

In [48]:
comparison = pd.DataFrame({
    'Original': original_stats,
    'Mean': df_mean.describe(),
    'Median': df_median.describe()
})
print(comparison)

         Original        Mean      Median
count  714.000000  891.000000  891.000000
mean    29.699118   29.699118   29.361582
std     14.526497   13.002015   13.019697
min      0.420000    0.420000    0.420000
25%     20.125000   22.000000   22.000000
50%     28.000000   29.699118   28.000000
75%     38.000000   35.000000   35.000000
max     80.000000   80.000000   80.000000


From the output analysis, we can see that Median is the more appropriate strategy. 

The Titanic age distribution contains outliers (seniors) that skew the mean. It is right-skewed. Mean is sensitive to these outliers while Median is robust. 
Using the median (the robust middle value) ensures that our imputed data remains representative of the majority of passengers without being pulled upward by a small number of elderly travelers. 
This aligns with the principles of handling non-homogeneous data. 

4) Apply hierarchical indexing on the Titanic DataFrame using 'Pclass' and 'Sex' as a multi-level index. Examine survival counts at each hierarchicallevel using xs() and loc[].

In [50]:
df_hindex = df.set_index(['Pclass', 'Sex']).sort_index()
df_hindex

PassengerId  Survived  \
Pclass Sex                             
1      female            2         1   
       female            4         1   
       female           12         1   
       female           32         1   
       female           53         1   
...                    ...       ...   
3      male            878         0   
       male            879         0   
       male            882         0   
       male            885         0   
       male            891         0   

                                                            Name   Age  SibSp  \
Pclass Sex                                                                      
1      female  Cumings, Mrs. John Bradley (Florence Briggs Th...  38.0      1   
       female       Futrelle, Mrs. Jacques Heath (Lily May Peel)  35.0      1   
       female                           Bonnell, Miss. Elizabeth  58.0      0   
       female     Spencer, Mrs. William Augustus (Marie Eugenie)   NaN      1   
       female           Harper, Mrs. Henry Sleeper (Myna Haxtun)  49.0      1   
...                                                          ...   ...    ...   
3      male                                 Petroff, Mr. Nedelio  19.0      0   
       male                                   Laleff, Mr. Kristo   NaN      0   
       male                                   Markun, Mr. Johann  33.0      0   
       male                               Sutehall, Mr. Henry Jr  25.0      0   
       male                                  Dooley, Mr. Patrick  32.0      0   

               Parch           Ticket      Fare Cabin Embarked AgeGroup  
Pclass Sex                                                               
1      female      0         PC 17599   71.2833   C85        C    Adult  
       female      0           113803   53.1000  C123        S    Adult  
       female      0           113783   26.5500  C103        S    Adult  
       female      0         PC 17569  146.5208   B78        C      NaN  
       female      0         PC 17572   76.7292   D33        C    Adult  
...              ...              ...       ...   ...      ...      ...  
3      male        0           349212    7.8958   NaN        S    Adult  
       male        0           349217    7.8958   NaN        S      NaN  
       male        0           349257    7.8958   NaN        S    Adult  
       male        0  SOTON/OQ 392076    7.0500   NaN        S    Adult  
       male        0           370376    7.7500   NaN        Q    Adult  

[891 rows x 11 columns]

In [51]:
df_hindex.xs('female', level='Sex')

,PassengerId,Survived,Name,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked,AgeGroup
Pclass,,,,,,,,,,,
1,2,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",38.0,1,0,PC 17599,71.2833,C85,C,Adult
1,4,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",35.0,1,0,113803,53.1000,C123,S,Adult
1,12,1,"Bonnell, Miss. Elizabeth",58.0,0,0,113783,26.5500,C103,S,Adult
1,32,1,"Spencer, Mrs. William Augustus (Marie Eugenie)",NaN,1,0,PC 17569,146.5208,B78,C,NaN
1,53,1,"Harper, Mrs. Henry Sleeper (Myna Haxtun)",49.0,1,0,PC 17572,76.7292,D33,C,Adult
...,...,...,...,...,...,...,...,...,...,...,...
3,864,0,"Sage, Miss. Dorothy Edith ""Dolly""",NaN,8,2,CA. 2343,69.5500,NaN,S,NaN
3,876,1,"Najib, Miss. Adele Kiamie ""Jane""",15.0,0,0,2667,7.2250,NaN,C,Teen
3,883,0,"Dahlberg, Miss. Gerda Ulrika",22.0,0,0,7552,10.5167,NaN,S,Adult


We are using xs() to return all the females survived across all Pclasses.
xs() is label-based, highly flexible, can handle unsorted data and is commonly used in comparing one category across all others.

In [52]:
df_hindex.loc[(1, 'male')] 

PassengerId  Survived                                  Name  \
Pclass Sex                                                                 
1      male            7         0               McCarthy, Mr. Timothy J   
       male           24         1          Sloper, Mr. William Thompson   
       male           28         0        Fortune, Mr. Charles Alexander   
       male           31         0              Uruchurtu, Don. Manuel E   
       male           35         0               Meyer, Mr. Edgar Joseph   
...                  ...       ...                                   ...   
       male          840         1                  Marechal, Mr. Pierre   
       male          858         1                Daly, Mr. Peter Denis    
       male          868         0  Roebling, Mr. Washington Augustus II   
       male          873         0              Carlsson, Mr. Frans Olof   
       male          890         1                 Behr, Mr. Karl Howell   

              Age  SibSp  Parch    Ticket      Fare        Cabin Embarked  \
Pclass Sex                                                                  
1      male  54.0      0      0     17463   51.8625          E46        S   
       male  28.0      0      0    113788   35.5000           A6        S   
       male  19.0      3      2     19950  263.0000  C23 C25 C27        S   
       male  40.0      0      0  PC 17601   27.7208          NaN        C   
       male  28.0      1      0  PC 17604   82.1708          NaN        C   
...           ...    ...    ...       ...       ...          ...      ...   
       male   NaN      0      0     11774   29.7000          C47        C   
       male  51.0      0      0    113055   26.5500          E17        S   
       male  31.0      0      0  PC 17590   50.4958          A24        S   
       male  33.0      0      0       695    5.0000  B51 B53 B55        S   
       male  26.0      0      0    111369   30.0000         C148        C   

            AgeGroup  
Pclass Sex            
1      male    Adult  
       male    Adult  
       male    Adult  
       male    Adult  
       male    Adult  
...              ...  
       male      NaN  
       male    Adult  
       male    Adult  
       male    Adult  
       male    Adult  

[122 rows x 11 columns]

We are using loc() to return a Dataframe of 1st class males. 
loc() is tuple-based, rigid, requires full path and is commonly used in getting a specific subgroup.

5) Create a new derived column 'FamilySize'=SibSp + Parch + 1. Using ufuncs, compute a 'SurvivalScore' that weighs Fare and FamilySize. Construct a final ranked DataFrame sorted by SurvivalScore.

i) Creating a new derived column

In [54]:
import numpy as np
df['FamilySize'] = df['SibSp'] + df['Parch'] + 1

ii) Using ufuncs and constructing final dataframe

In [57]:
df['SurvivalScore'] = np.log1p(df['Fare']) / df['FamilySize']
ranked_df = df.sort_values(by='SurvivalScore', ascending=False)
ranked_df

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked,AgeGroup,FamilySize,SurvivalScore
258,259,1,1,"Ward, Miss. Anna",female,35.0,0,0,PC 17755,512.3292,NaN,C,Adult,1,6.240917
737,738,1,1,"Lesurer, Mr. Gustave J",male,35.0,0,0,PC 17755,512.3292,B101,C,Adult,1,6.240917
557,558,0,1,"Robbins, Mr. Victor",male,NaN,0,0,PC 17757,227.5250,NaN,C,NaN,1,5.431646
380,381,1,1,"Bidois, Miss. Rosalie",female,42.0,0,0,PC 17757,227.5250,NaN,C,Adult,1,5.431646
716,717,1,1,"Endres, Miss. Caroline Louise",female,38.0,0,0,PC 17757,227.5250,C45,C,Adult,1,5.431646
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
179,180,0,3,"Leonard, Mr. Lionel",male,36.0,0,0,LINE,0.0000,NaN,S,Adult,1,0.000000
597,598,0,3,"Johnson, Mr. Alfred",male,49.0,0,0,LINE,0.0000,NaN,S,Adult,1,0.000000
466,467,0,2,"Campbell, Mr. William",male,NaN,0,0,239853,0.0000,NaN,S,NaN,1,0.000000
633,634,0,1,"Parr, Mr. William Henry Marsh",male,NaN,0,0,112052,0.0000,NaN,S,NaN,1,0.000000


In [56]:
print(ranked_df[['Name', 'Fare', 'FamilySize', 'SurvivalScore']].head())

                              Name      Fare  FamilySize  SurvivalScore
258               Ward, Miss. Anna  512.3292           1       6.240917
737         Lesurer, Mr. Gustave J  512.3292           1       6.240917
557            Robbins, Mr. Victor  227.5250           1       5.431646
380          Bidois, Miss. Rosalie  227.5250           1       5.431646
716  Endres, Miss. Caroline Louise  227.5250           1       5.431646


Simply dividing Fare by FamilySize (Fare-per-person) is a logical way to "weigh" the two variables. However, because Fare has extreme outliers (some paid 0 dollars, others paid 512 dollars), using np.log1p is superior. It squashes the range so that a few wealthy passengers don't completely break the ranking system.

It handles the skewed distribution of fares and accounts for the fact that larger families faced more survival friction.

By using df.sort_values(), a 'Priority List' is created. Final output shows that single passengers in 1st Class appear at the very top, as they had high fares and a family size of 1, resulting in the highest scores.

These operations are vectorized. This means Pandas performs the calculation for all 891 rows simultaneously at the C-level, rather than using a slow Python loop.

6) Perform index alignment between two Series: one containing average fare per Pclass, and anothercontaining average age per Pclass. Examine how Pandas aligns indices during arithmetic operations.

In [58]:
avg_fare = df.groupby('Pclass')['Fare'].mean()
avg_age = df.groupby('Pclass')['Age'].mean()

In [61]:
avg_fare

Pclass
1    84.154687
2    20.662183
3    13.675550
Name: Fare, dtype: float64

In [62]:
avg_age

Pclass
1    38.233441
2    29.877630
3    25.140620
Name: Age, dtype: float64

In [60]:
avg_age_reordered = avg_age.reindex([3, 1, 2]) 
avg_age_reordered

Pclass
3    25.140620
1    38.233441
2    29.877630
Name: Age, dtype: float64

To demonstrate alignment clearly, manually reordering one will show the difference clearly.

In [63]:
addition = avg_fare + avg_age_reordered
subtraction = avg_fare - avg_age_reordered
multiplication = avg_fare * avg_age_reordered
division = avg_fare / avg_age_reordered

In [65]:
addition

Pclass
1    122.388128
2     50.539813
3     38.816170
dtype: float64

In [66]:
subtraction

Pclass
1    45.921247
2    -9.215447
3   -11.465070
dtype: float64

In [67]:
multiplication

Pclass
1    3217.523268
2     617.337064
3     343.811805
dtype: float64

In [68]:
division

Pclass
1    2.201075
2    0.691560
3    0.543962
dtype: float64

In [69]:
alignment_check = pd.DataFrame({
    'Avg Fare': avg_fare,
    'Avg Age (Reordered)': avg_age_reordered,
    'Addition': addition,
    'Division': division,
})

print(alignment_check)

         Avg Fare  Avg Age (Reordered)    Addition  Division
Pclass                                                      
1       84.154687            38.233441  122.388128  2.201075
2       20.662183            29.877630   50.539813  0.691560
3       13.675550            25.140620   38.816170  0.543962


Pandas does not perform arithmetic operations based on the position of the data (like a list or NumPy array), but based on the index labels.

Even though we reordered avg_age_reordered to start with Pclass 3, Pandas ignores the physical order. 
When avg_fare + avg_age_reordered is executed, it looks for the label 1 in both Series, adds those specific values, and places the result at index 1.
Even when the data is "jumbled", Pandas correctly identifies that the Fare for Pclass 3 should only be added to the Age for Pclass 3.

7) Examine the relationship between 'Embarked' port and survival rate using groupby with hierarchical indexing. Assess whether port of embarkation significantly influenced survival outcomes.

In [70]:
embarked_survival_rates = df.groupby(['Embarked', 'Pclass'])['Survived'].mean()
embarked_survival_rates

Embarked  Pclass
C         1         0.694118
          2         0.529412
          3         0.378788
Q         1         0.500000
          2         0.666667
          3         0.375000
S         1         0.582677
          2         0.463415
          3         0.189802
Name: Survived, dtype: float64

This provides the Rate requested and uses hierarchical indexing to provide a deeper assessment of the class-port relationship.

Passengers who embarked at Cherbourg (C) consistently show higher survival rates across almost all classes compared to Southampton (S).
While the port appears to have an influence, hierarchical indexing reveals that it is strongly correlated with Pclass. A much higher proportion of Cherbourg passengers were in 1st Class.

The port of embarkation influenced survival outcomes primarily because it served as a proxy for socioeconomic status (Pclass). However, even within the same class, Cherbourg survival rates often remain slightly higher, suggesting secondary factors like cabin location on the ship based on boarding order.

8) Design a function using apply() and map() to categorize passengers into 'Child', 'Adult', and 'Senior'; based on age. Integrate this with a pivot table showing survival rates per age category and gender.

i) Categorizing passengers

In [71]:
def categorize_age(age):
    if pd.isnull(age):
        return 'Unknown'
    if age < 18:
        return 'Child'
    elif age < 60:
        return 'Adult'
    else:
        return 'Senior'

ii) Using apply() and map() functions

In [72]:
df['AgeCategory'] = df['Age'].apply(categorize_age)

In [73]:
df['SurvivalStatus'] = df['Survived'].map({0: 'Died', 1: 'Survived'})

In Pandas, apply() is typically used for applying a custom function along an axis (like a column), while map() is often used for element-wise transformations or substituting values based on a dictionary/Series.
We use apply() to execute the custom function logic on the 'Age' column.
We use map() to create a human-readable 'Survived' label.

iii) Integrating with a pivot table

In [74]:
pivot_table = df.pivot_table(values='Survived', 
                             index='AgeCategory', 
                             columns='Sex', 
                             aggfunc='mean')

pivot_table

Sex,female,male
AgeCategory,,
Adult,0.767327,0.179625
Child,0.690909,0.396552
Senior,1.000000,0.136364
Unknown,0.679245,0.129032


From the pivot table, we can see that Adult Females and Children have the highest survival rates, while Adult Males have the lowest. 
This statistical evidence supports the historical priority given to vulnerable groups during the evacuation.

9) Examine null value patterns across all columns. Using isnull(), notnull(), and dropna() with different thresh and axis parameters, determine the optimal strategy for cleaning the dataset.

In [75]:
df.isnull().sum()

PassengerId         0
Survived            0
Pclass              0
Name                0
Sex                 0
Age               177
SibSp               0
Parch               0
Ticket              0
Fare                0
Cabin             687
Embarked            2
AgeGroup          177
FamilySize          0
SurvivalScore       0
AgeCategory         0
SurvivalStatus      0
dtype: int64

In [76]:
df.notnull().sum()

PassengerId       891
Survived          891
Pclass            891
Name              891
Sex               891
Age               714
SibSp             891
Parch             891
Ticket            891
Fare              891
Cabin             204
Embarked          889
AgeGroup          714
FamilySize        891
SurvivalScore     891
AgeCategory       891
SurvivalStatus    891
dtype: int64

In [77]:
df.dropna(axis=1, thresh=700)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Embarked,AgeGroup,FamilySize,SurvivalScore,AgeCategory,SurvivalStatus
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,S,Adult,2,1.055107,Adult,Died
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C,Adult,2,2.140297,Adult,Survived
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,S,Adult,1,2.188856,Adult,Survived
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,S,Adult,2,1.995417,Adult,Survived
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,S,Adult,1,2.202765,Adult,Died
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.0000,S,Adult,1,2.639057,Adult,Died
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.0000,S,Adult,1,3.433987,Adult,Survived
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,NaN,1,2,W./C. 6607,23.4500,S,NaN,4,0.799158,Unknown,Died
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.0000,C,Adult,1,3.433987,Adult,Survived


We use this for columns like 'Cabin'.
If a column doesn't meet a minimum requirement of "real" data, the entire feature is unreliable for analysis or machine learning. 
Dropping the column is better than guessing 77% of the values.

In [78]:
df.dropna(axis=0, thresh=5)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked,AgeGroup,FamilySize,SurvivalScore,AgeCategory,SurvivalStatus
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S,Adult,2,1.055107,Adult,Died
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C,Adult,2,2.140297,Adult,Survived
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S,Adult,1,2.188856,Adult,Survived
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S,Adult,2,1.995417,Adult,Survived
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S,Adult,1,2.202765,Adult,Died
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.0000,NaN,S,Adult,1,2.639057,Adult,Died
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.0000,B42,S,Adult,1,3.433987,Adult,Survived
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,NaN,1,2,W./C. 6607,23.4500,NaN,S,NaN,4,0.799158,Unknown,Died
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.0000,C148,C,Adult,1,3.433987,Adult,Survived


We use this as a safety net for the entire DataFrame.
This ensures that if a specific passenger record is almost entirely empty (e.g., it only has a Name and ID but no Class, Age, or Survival data), it gets removed. 
A threshold of 5 is conservative as it keeps almost everyone except the most broken records.

In [80]:
df_optimal = df.drop(columns=['Cabin'])
df_optimal['Age'] = df_optimal['Age'].fillna(df_optimal['Age'].median())
df_optimal

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Embarked,AgeGroup,FamilySize,SurvivalScore,AgeCategory,SurvivalStatus
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,S,Adult,2,1.055107,Adult,Died
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C,Adult,2,2.140297,Adult,Survived
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,S,Adult,1,2.188856,Adult,Survived
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,S,Adult,2,1.995417,Adult,Survived
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,S,Adult,1,2.202765,Adult,Died
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.0000,S,Adult,1,2.639057,Adult,Died
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.0000,S,Adult,1,3.433987,Adult,Survived
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,28.0,1,2,W./C. 6607,23.4500,S,NaN,4,0.799158,Unknown,Died
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.0000,C,Adult,1,3.433987,Adult,Survived


In [81]:
df_final = df_optimal.dropna(subset=['Embarked'])
df_final

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Embarked,AgeGroup,FamilySize,SurvivalScore,AgeCategory,SurvivalStatus
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,S,Adult,2,1.055107,Adult,Died
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C,Adult,2,2.140297,Adult,Survived
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,S,Adult,1,2.188856,Adult,Survived
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,S,Adult,2,1.995417,Adult,Survived
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,S,Adult,1,2.202765,Adult,Died
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.0000,S,Adult,1,2.639057,Adult,Died
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.0000,S,Adult,1,3.433987,Adult,Survived
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,28.0,1,2,W./C. 6607,23.4500,S,NaN,4,0.799158,Unknown,Died
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.0000,C,Adult,1,3.433987,Adult,Survived


Based on 'Handling Missing Values' principles, df_optimal and df_final are the most optimal because they follow the three main pillars of data cleaning:

Drop Columns when the data is too sparse to be trusted (Cabin).
Impute values when the data is essential but partially missing (Age).
Drop Rows when the missingness is so small that removal doesn't hurt the sample size (Embarked).

10) Create a comprehensive summary DataFrame using multi-level groupby combining Pclass, Sex, and Survived. Use stack() and unstack() to reshape the result into a cross-tabulation report. Upload the complete notebook to your GitHub repository.

i) Comprehensive summary

In [83]:
summary = df.groupby(['Pclass', 'Sex', 'Survived']).size()
summary

Pclass  Sex     Survived
1       female  0             3
                1            91
        male    0            77
                1            45
2       female  0             6
                1            70
        male    0            91
                1            17
3       female  0            72
                1            72
        male    0           300
                1            47
dtype: int64

This counts the number of passengers for every Class-Sex-Survival combination.

ii) Using stack() and unstack() functions

In [85]:
cross_tab = summary.unstack()
final_report = cross_tab.fillna(0).astype(int)
restacked_report = final_report.stack()
final_report

Survived         0   1
Pclass Sex            
1      female    3  91
       male     77  45
2      female    6  70
       male     91  17
3      female   72  72
       male    300  47

unstack() moves the 'Survived' level (the innermost level) to the columns.
To handle missing values, fillna(0) is used and ensures that the report is complete and professional.

In [86]:
restacked_report

Pclass  Sex     Survived
1       female  0             3
                1            91
        male    0            77
                1            45
2       female  0             6
                1            70
        male    0            91
                1            17
3       female  0            72
                1            72
        male    0           300
                1            47
dtype: int32

This moves the columns back into the row index. 